Include/Rover/TrialView.hpp

```diff
@@ -108,40 +118,37 @@ namespace Rover {
       const Sample& operator [](std::size_t index) const;
 
     private:
+      std::size_t m_size;
+      GetPtr m_get_ptr;
       ConstIterator m_begin;
       ConstIterator m_end;
-      std::size_t m_size;
   };
 
   template<typename Trial>
   TrialView(const Trial&) -> TrialView<typename Trial::Sample>;
 
   template<typename S>
-  template<typename Trial>
-  TrialView<S>::ConstIterator::ConstIterator(const Trial& t,
-      std::ptrdiff_t offset)
-    : m_offset(offset),
-      m_get_ptr([begin = t.begin()] (std::ptrdiff_t offset)
-      {
-        return &(*(begin + offset));
-      }) {}
+  TrialView<S>::ConstIterator::ConstIterator(const GetPtr& get_ptr,
+      std::size_t offset)
+    : m_get_ptr(&get_ptr),
+      m_offset(static_cast<std::ptrdiff_t>(offset)) {}
```

Comment:
Use a ```std::size_t``` for ```m_offset```.

Reply:
But why? ptrdiff_t is the native pointer/iterator arithmetic type, and arithmetics for ListTrial<S>::ConstIterator = std::vector<S>::const_iterator work with that type. If I end up passing size_t to its operator+ I will be making use of the implicit conversion each time. Also, arithmetic operators for TrialView<S>::ConstIterator must work with ptrdiff, as otherwise
auto i = -1;
it += i;
would not work. Since m_offset is an implementation detail, it should be chosen according to convenience and efficiency, and, in these terms, having one cast in ConstIterator constructor seems better than doing the opposite cast on every operator call.
Even conceptually, m_offset should be signed as
auto it = v.begin();
it -= 10;
it += 15;
is a valid syntax. Not saying it won't work with size_t, but just trying to understand the rationale for making it unsigned. Out-of-bound iterators cannot be dereferenced but they still exist.

Reply:
No idea why it's crossing it out.

Reply:
```std::ptrdiff_t``` is the native arithmetic type only when negative values are possible, otherwise ```std::size_t``` is the native arithmetic type. There is no valid use case for a negative index.

As for the use case:

```
auto i = -1;
it += i;
```

That will work even for ```std::size_t``` in that the result will decrement ```it``` and that behavior is guaranteed to work. As for out-of-bound iterators, the only valid out-of-bound iterator is one-past-the-end. Any other out of bound iterator is invalid and should not be considered.

Reply:
Well, std::vector uses ptrdiff_t and it does not allow for negative indexes.
I think it has to do with algebraic symmetry, i.e.
it1 - it2 == x, x must be signed as it1 < it2 is possible
RandomAccessIterator concept requires operator- to be consistent with operator+, so the above equation is identical to
it1 - x == it2 and it2 + x == it1 with all the same types. It, in turn, leads to (it1 -= x) == it2 as well and so on.
I am not saying it won't work with size_t, it will. Just trying to understand why it should be such.

Reply:
I am not sure if what the Standard means by iterator validity has to do something with bounds. As far as I know, iterators are only invalid once one of iterator invalidation operation is executed, such as std::vector re-allocation.
I also think RandomAccessIterator is supposed to define a abelian group over subtraction, even though I cannot find the exact wording for that. When there is such a group, out-of-bound operators are possible, for example, by chaining the following:
auto offset = end - begin;
auto it = end + offset;

Reply:
```std::vector``` does not use ```std::ptrdiff_t```, it uses an unsigned type that is implementation defined but most commonly ```std::size_t```. While the Standard does not explicitly say that it is undefined behavior to create an iterator that is out of bounds, the standard does state in section 24.2.7 that the operation ```++i``` for a random access iterator ```i``` must be dereferenceable and by definition the operation ```i + x``` for some constant x has the same operational semantics as ```++(++(++(...x times...)i))```. Similar rules apply for ```--i``` and ```i - x```. As such the construction of an iterator that is out-of-bounds and not one-past-the-end is undefined behavior in all cases.